### ISSUE TITLE
cmus-status-scrobbler can't scrobble any further

### ISSUE THREAD

**Issue by GUUser91** [vjeranc/cmus-status-scrobbler#5]
I downloaded this album linked below. I converted it to flac. After scrobbling the 41st track of this same album in flac, it stops scrobbling.
http://snesmusic.org/v2/profile.php?profile=set&selected=783

```
INFO:root:Status(status='playing', file='/home/bud/Audio/FLAC-Earthbound (1995-06-05) (SNES)/041 - Battle Against a Weak Opponent.flac', artist='田中宏和, 鈴木慶一 & 金津宏 ', albumartist=None, album='Earthbound', discnumber=1, tracknumber='041', title='Battle Against a Weak Opponent', date='1995-06-05', duration='58', musicbrainz_trackid=None, cur_time=datetime.datetime(2023, 1, 8, 13, 29, 26, 644826))
INFO:root:Sending now playing for lastfm
INFO:root:{'artist': '田中宏和, 鈴木慶一 & 金津宏 ', 'track': 'Battle Against a Weak Opponent', 'album': 'Earthbound', 'trackNumber': '041', 'duration': '58', 'sk': 'IiLYEx0OvvLdlVs0r-JevMVNkuVlBB5O', 'api_key': 'b20a1f33c3b3b105de2246a54624b6a4', 'method': 'track.updateNowPlaying', 'api_sig': '4694d76fc8d3c413c89f69a8bcf947b3', 'format': 'json'}
INFO:root:{"nowplaying":{"artist":{"corrected":"0","#text":"田中宏和, 鈴木慶一 & 金津宏"},"track":{"corrected":"0","#text":"Battle Against a Weak Opponent"},"ignoredMessage":{"code":"0","#text":""},"albumArtist":{"corrected":"0","#text":""},"album":{"corrected":"0","#text":"Earthbound"}}}
INFO:root:Scrobbling previous tracks for lastfm
INFO:root:{'sk': 'IiLYEx0OvvLdlVs0r-JevMVNkuVlBB5O', 'artist[0]': '田中宏和, 鈴木慶一 & 金津宏 ', 'track[0]': 'Ambush!', 'timestamp[0]': '1673184562', 'album[0]': 'Earthbound', 'trackNumber[0]': '040', 'duration[0]': '4', 'api_key': 'b20a1f33c3b3b105de2246a54624b6a4', 'method': 'track.scrobble', 'api_sig': '2033d232b1f5c47136792403a9bf7e2e', 'format': 'json'}
INFO:root:{"scrobbles":{"scrobble":{"artist":{"corrected":"0","#text":"田中宏和, 鈴木慶一 & 金津宏"},"album":{"corrected":"0","#text":"Earthbound"},"track":{"corrected":"0","#text":"Ambush!"},"ignoredMessage":{"code":"1","#text":""},"albumArtist":{"corrected":"0","#text":""},"timestamp":"1673184562"},"@attr":{"ignored":1,"accepted":0}}}
INFO:root:Sending now playing for librefm
INFO:root:{'artist': '田中宏和, 鈴木慶一 & 金津宏 ', 'track': 'Battle Against a Weak Opponent', 'album': 'Earthbound', 'trackNumber': '041', 'duration': '58', 'sk': '51634aa8d9625e348778d072c082eee4', 'api_key': 'b20a1f33c3b3b105de2246a54624b6a4', 'method': 'track.updateNowPlaying', 'api_sig': 'cbf0564b5483fa5c12d6ea4e062f464c', 'format': 'json'}
INFO:root:{"nowplaying":{"track":{"#text":"Battle Against a Weak Opponent","corrected":"0"},"artist":{"#text":"\u7530\u4e2d\u5b8f\u548c, \u9234\u6728\u6176\u4e00 & \u91d1\u6d25\u5b8f","corrected":"1"},"album":{"#text":"Earthbound","corrected":"0"},"albumArtist":{"#text":"","corrected":"0"},"ignoredMessage":{"#text":"","code":"0"}}}
INFO:root:Scrobbling previous tracks for librefm
INFO:root:{'sk': '51634aa8d9625e348778d072c082eee4', 'artist[0]': '田中宏和, 鈴木慶一 & 金津宏 ', 'track[0]': 'Ambush!', 'timestamp[0]': '1673184562', 'album[0]': 'Earthbound', 'trackNumber[0]': '040', 'duration[0]': '4', 'api_key': 'b20a1f33c3b3b105de2246a54624b6a4', 'method': 'track.scrobble', 'api_sig': 'b6e4ab2c0596712ec5028e474333792e', 'format': 'json'}
INFO:root:{"scrobbles":{"scrobble":{"track":{"#text":"Ambush!","corrected":"0"},"artist":{"#text":"\u7530\u4e2d\u5b8f\u548c, \u9234\u6728\u6176\u4e00 & \u91d1\u6d25\u5b8f","corrected":"1"},"album":{"#text":"Earthbound","corrected":"0"},"albumArtist":{"#text":"","corrected":"0"},"timestamp":"1673184562","ignoredMessage":{"#text":"","code":"0"}},"@attr":{"accepted":"1","ignored":"0"}}}
INFO:root:Scrobbling previous tracks for listenbrainz
INFO:root:{'sk': '9e8f8d94edd8225d0b41847bf7d200e8c72b2578', 'artist[0]': '田中宏和, 鈴木慶一 & 金津宏 ', 'track[0]': 'Ambush!', 'timestamp[0]': '1673184562', 'album[0]': 'Earthbound', 'trackNumber[0]': '040', 'duration[0]': '4', 'api_key': 'b20a1f33c3b3b105de2246a54624b6a4', 'method': 'track.scrobble', 'api_sig': 'b5590fd0617214afb8e2e60c1a00e793'}
INFO:root:<?xml version="1.0" encoding="utf-8"?>
<lfm status="ok">
  <scrobbles accepted="1" ignored="0">
    <scrobble>
      <track corrected="0">Ambush!</track>
      <artist corrected="1">田中宏和, 鈴木慶一 &amp; 金津宏</artist>
      <album corrected="0">Earthbound</album>
      <albumArtist corrected="1">田中宏和, 鈴木慶一 &amp; 金津宏</albumArtist>
      <timestamp>1673184562</timestamp>
      <ignoredMessage code="0"></ignoredMessage>
    </scrobble>
  </scrobbles>
</lfm>
INFO:root:Status(status='paused', file='/home/bud/Audio/FLAC-Earthbound (1995-06-05) (SNES)/041 - Battle Against a Weak Opponent.flac', artist='田中宏和, 鈴木慶一 & 金津宏 ', albumartist=None, album='Earthbound', discnumber=1, tracknumber='041', title='Battle Against a Weak Opponent', date='1995-06-05', duration='58', musicbrainz_trackid=None, cur_time=datetime.datetime(2023, 1, 8, 13, 30, 17, 659220))
INFO:root:Status(status='playing', file='/home/bud/Audio/FLAC-Earthbound (1995-06-05) (SNES)/041 - Battle Against a Weak Opponent.flac', artist='田中宏和, 鈴木慶一 & 金津宏 ', albumartist=None, album='Earthbound', discnumber=1, tracknumber='041', title='Battle Against a Weak Opponent', date='1995-06-05', duration='58', musicbrainz_trackid=None, cur_time=datetime.datetime(2023, 1, 8, 13, 30, 23, 52205))
INFO:root:Sending now playing for lastfm
INFO:root:{'artist': '田中宏和, 鈴木慶一 & 金津宏 ', 'track': 'Battle Against a Weak Opponent', 'album': 'Earthbound', 'trackNumber': '041', 'duration': '58', 'sk': 'IiLYEx0OvvLdlVs0r-JevMVNkuVlBB5O', 'api_key': 'b20a1f33c3b3b105de2246a54624b6a4', 'method': 'track.updateNowPlaying', 'api_sig': '4694d76fc8d3c413c89f69a8bcf947b3', 'format': 'json'}
INFO:root:{"nowplaying":{"artist":{"corrected":"0","#text":"田中宏和, 鈴木慶一 & 金津宏"},"track":{"corrected":"0","#text":"Battle Against a Weak Opponent"},"ignoredMessage":{"code":"0","#text":""},"albumArtist":{"corrected":"0","#text":""},"album":{"corrected":"0","#text":"Earthbound"}}}
INFO:root:Sending now playing for librefm
INFO:root:{'artist': '田中宏和, 鈴木慶一 & 金津宏 ', 'track': 'Battle Against a Weak Opponent', 'album': 'Earthbound', 'trackNumber': '041', 'duration': '58', 'sk': '51634aa8d9625e348778d072c082eee4', 'api_key': 'b20a1f33c3b3b105de2246a54624b6a4', 'method': 'track.updateNowPlaying', 'api_sig': 'cbf0564b5483fa5c12d6ea4e062f464c', 'format': 'json'}
INFO:root:{"nowplaying":{"track":{"#text":"Battle Against a Weak Opponent","corrected":"0"},"artist":{"#text":"\u7530\u4e2d\u5b8f\u548c, \u9234\u6728\u6176\u4e00 & \u91d1\u6d25\u5b8f","corrected":"1"},"album":{"#text":"Earthbound","corrected":"0"},"albumArtist":{"#text":"","corrected":"0"},"ignoredMessage":{"#text":"","code":"0"}}}
INFO:root:Status(status='playing', file='/home/bud/Audio/FLAC-Earthbound (1995-06-05) (SNES)/042 - Battle Against a Weird Opponent.flac', artist='田中宏和, 鈴木慶一 & 金津宏 ', albumartist=None, album='Earthbound', discnumber=1, tracknumber='042', title='Battle Against a Weird Opponent', date='1995-06-05', duration='75', musicbrainz_trackid=None, cur_time=datetime.datetime(2023, 1, 8, 13, 30, 29, 795392))
INFO:root:Sending now playing for lastfm
INFO:root:{'artist': '田中宏和, 鈴木慶一 & 金津宏 ', 'track': 'Battle Against a Weird Opponent', 'album': 'Earthbound', 'trackNumber': '042', 'duration': '75', 'sk': 'IiLYEx0OvvLdlVs0r-JevMVNkuVlBB5O', 'api_key': 'b20a1f33c3b3b105de2246a54624b6a4', 'method': 'track.updateNowPlaying', 'api_sig': '79c889175e6f83514fbb6b4b9eb7b3f8', 'format': 'json'}
INFO:root:{"nowplaying":{"artist":{"corrected":"0","#text":"田中宏和, 鈴木慶一 & 金津宏"},"track":{"corrected":"0","#text":"Battle Against a Weird Opponent"},"ignoredMessage":{"code":"0","#text":""},"albumArtist":{"corrected":"0","#text":""},"album":{"corrected":"0","#text":"Earthbound"}}}
INFO:root:Scrobbling previous tracks for lastfm
INFO:root:{'sk': 'IiLYEx0OvvLdlVs0r-JevMVNkuVlBB5O', 'artist[0]': '田中宏和, 鈴木慶一 & 金津宏 ', 'track[0]': 'Battle Against a Weak Opponent', 'timestamp[0]': '1673184623', 'album[0]': 'Earthbound', 'trackNumber[0]': '041', 'duration[0]': '58', 'api_key': 'b20a1f33c3b3b105de2246a54624b6a4', 'method': 'track.scrobble', 'api_sig': '2233e70dcd037cf57ed2cbaf6c12779a', 'format': 'json'}
INFO:root:{"scrobbles":{"scrobble":{"artist":{"corrected":"0","#text":"田中宏和, 鈴木慶一 & 金津宏"},"album":{"corrected":"0","#text":"Earthbound"},"track":{"corrected":"0","#text":"Battle Against a Weak Opponent"},"ignoredMessage":{"code":"0","#text":""},"albumArtist":{"corrected":"0","#text":""},"timestamp":"1673184623"},"@attr":{"ignored":0,"accepted":1}}}
INFO:root:Sending now playing for librefm
INFO:root:{'artist': '田中宏和, 鈴木慶一 & 金津宏 ', 'track': 'Battle Against a Weird Opponent', 'album': 'Earthbound', 'trackNumber': '042', 'duration': '75', 'sk': '51634aa8d9625e348778d072c082eee4', 'api_key': 'b20a1f33c3b3b105de2246a54624b6a4', 'method': 'track.updateNowPlaying', 'api_sig': 'e5f5f7d1e1598e55166bd43cb099d802', 'format': 'json'}
ERROR:root:Error happened
Traceback (most recent call last):
  File "/home/bud/.config/cmus/cmus_status_scrobbler.py", line 443, in <module>
    main()
  File "/home/bud/.config/cmus/cmus_status_scrobbler.py", line 432, in main
    with db_connect(conf['global'].get('db_path', args.db_path),
  File "/home/bud/.config/cmus/cmus_status_scrobbler.py", line 372, in db_connect
    con.execute('BEGIN IMMEDIATE')  # synchronizing processes
sqlite3.OperationalError: database is locked
ERROR:root:Error happened
Traceback (most recent call last):
  File "/home/bud/.config/cmus/cmus_status_scrobbler.py", line 443, in <module>
    main()
  File "/home/bud/.config/cmus/cmus_status_scrobbler.py", line 432, in main
    with db_connect(conf['global'].get('db_path', args.db_path),
  File "/home/bud/.config/cmus/cmus_status_scrobbler.py", line 372, in db_connect
    con.execute('BEGIN IMMEDIATE')  # synchronizing processes
sqlite3.OperationalError: database is locked
ERROR:root:Error happened
Traceback (most recent call last):
  File "/home/bud/.config/cmus/cmus_status_scrobbler.py", line 443, in <module>
    main()
  File "/home/bud/.config/cmus/cmus_status_scrobbler.py", line 432, in main
    with db_connect(conf['global'].get('db_path', args.db_path),
  File "/home/bud/.config/cmus/cmus_status_scrobbler.py", line 372, in db_connect
    con.execute('BEGIN IMMEDIATE')  # synchronizing processes
sqlite3.OperationalError: database is locked
ERROR:root:Error happened
Traceback (most recent call last):
  File "/home/bud/.config/cmus/cmus_status_scrobbler.py", line 443, in <module>
    main()
  File "/home/bud/.config/cmus/cmus_status_scrobbler.py", line 432, in main
    with db_connect(conf['global'].get('db_path', args.db_path),
  File "/home/bud/.config/cmus/cmus_status_scrobbler.py", line 372, in db_connect
    con.execute('BEGIN IMMEDIATE')  # synchronizing processes
sqlite3.OperationalError: database is locked
ERROR:root:Error happened
Traceback (most recent call last):
  File "/home/bud/.config/cmus/cmus_status_scrobbler.py", line 443, in <module>
    main()
  File "/home/bud/.config/cmus/cmus_status_scrobbler.py", line 432, in main
    with db_connect(conf['global'].get('db_path', args.db_path),
  File "/home/bud/.config/cmus/cmus_status_scrobbler.py", line 372, in db_connect
    con.execute('BEGIN IMMEDIATE')  # synchronizing processes
sqlite3.OperationalError: database is locked
ERROR:root:Error happened
Traceback (most recent call last):
  File "/home/bud/.config/cmus/cmus_status_scrobbler.py", line 443, in <module>
    main()
  File "/home/bud/.config/cmus/cmus_status_scrobbler.py", line 432, in main
    with db_connect(conf['global'].get('db_path', args.db_path),
  File "/home/bud/.config/cmus/cmus_status_scrobbler.py", line 372, in db_connect
    con.execute('BEGIN IMMEDIATE')  # synchronizing processes
sqlite3.OperationalError: database is locked
ERROR:root:Error happened
Traceback (most recent call last):
  File "/home/bud/.config/cmus/cmus_status_scrobbler.py", line 443, in <module>
    main()
  File "/home/bud/.config/cmus/cmus_status_scrobbler.py", line 432, in main
    with db_connect(conf['global'].get('db_path', args.db_path),
  File "/home/bud/.config/cmus/cmus_status_scrobbler.py", line 372, in db_connect
    con.execute('BEGIN IMMEDIATE')  # synchronizing processes
sqlite3.OperationalError: database is locked
ERROR:root:Error happened
Traceback (most recent call last):
  File "/home/bud/.config/cmus/cmus_status_scrobbler.py", line 443, in <module>
    main()
  File "/home/bud/.config/cmus/cmus_status_scrobbler.py", line 432, in main
    with db_connect(conf['global'].get('db_path', args.db_path),
  File "/home/bud/.config/cmus/cmus_status_scrobbler.py", line 372, in db_connect
    con.execute('BEGIN IMMEDIATE')  # synchronizing processes
sqlite3.OperationalError: database is locked
ERROR:root:Error happened
Traceback (most recent call last):
  File "/home/bud/.config/cmus/cmus_status_scrobbler.py", line 443, in <module>
    main()
  File "/home/bud/.config/cmus/cmus_status_scrobbler.py", line 432, in main
    with db_connect(conf['global'].get('db_path', args.db_path),
  File "/home/bud/.config/cmus/cmus_status_scrobbler.py", line 372, in db_connect
    con.execute('BEGIN IMMEDIATE')  # synchronizing processes
sqlite3.OperationalError: database is locked
ERROR:root:Error happened
Traceback (most recent call last):
  File "/home/bud/.config/cmus/cmus_status_scrobbler.py", line 443, in <module>
    main()
  File "/home/bud/.config/cmus/cmus_status_scrobbler.py", line 432, in main
    with db_connect(conf['global'].get('db_path', args.db_path),
  File "/home/bud/.config/cmus/cmus_status_scrobbler.py", line 372, in db_connect
    con.execute('BEGIN IMMEDIATE')  # synchronizing processes
sqlite3.OperationalError: database is locked
```

**Comment by vjeranc** [comment]
Thanks for the log. It's very useful.

Which python and cmus version are you running?

I'll take some time today to see why exactly this lock happens. Might be that I need to close a cursor/connection at some point.

**Comment by GUUser91** [comment]
@vjeranc
cmus 2.10.0-3
Python 3.10.9

**Comment by vjeranc** [comment]
Thanks for reporting the error.

I pushed a few commits that should avoid the error.

`BEGIN IMMEDIATE` is a way to allow the process to wait on a locked DB and then continue, but looks like sqlite still in some cases returns an error for `BEGIN IMMEDIATE`.

Without this one can experience concurrency bugs when holding pause/play (sending a lot of pause play events in succession).

But turns out you landed on this SQLITE_BUSY error. Hopefully the retry I added will work.

I also moved `now_playing` notifications to happen outside of the DB lock. So this should improve it too.

Cheers!
